### PR TITLE
Add support for computed columns

### LINKED ISSUE #624: Computed column migration file generated incorrectly
SQLAlchemy version 1.3.11 (just released on November 11th, 2019) includes support for [Computed columns](https://docs.sqlalchemy.org/en/13/core/defaults.html).  From their documentation an example Computed column definition would be something like:

`Column("area", Integer, Computed("side * side"))`

After adding such a Computed column definition to my table model and trying to use alembic to autogenerate a migration file I encounter the following error:

```  File "/home/dev.internal/developer/sandbox/mbs/venv/lib/python3.7/site-packages/alembic/script/revision.py", line 321, in get_revisions
    resolved_id, branch_label = self._resolve_revision_number(id_)
  File "/home/dev.internal/developer/sandbox/mbs/venv/lib/python3.7/site-packages/alembic/script/revision.py", line 501, in _resolve_revision_number
    self._revision_map
  File "/home/dev.internal/developer/sandbox/mbs/venv/lib/python3.7/site-packages/alembic/util/langhelpers.py", line 230, in __get__
    obj.__dict__[self.__name__] = result = self.fget(obj)
  File "/home/dev.internal/developer/sandbox/mbs/venv/lib/python3.7/site-packages/alembic/script/revision.py", line 123, in _revision_map
    for revision in self._generator():
  File "/home/dev.internal/developer/sandbox/mbs/venv/lib/python3.7/site-packages/alembic/script/base.py", line 112, in _load_revisions
    script = Script._from_filename(self, vers, file_)
  File "/home/dev.internal/developer/sandbox/mbs/venv/lib/python3.7/site-packages/alembic/script/base.py", line 906, in _from_filename
    module = util.load_python_file(dir_, filename)
  File "/home/dev.internal/developer/sandbox/mbs/venv/lib/python3.7/site-packages/alembic/util/pyfiles.py", line 98, in load_python_file
    module = load_module_py(module_id, path)
  File "/home/dev.internal/developer/sandbox/mbs/venv/lib/python3.7/site-packages/alembic/util/compat.py", line 173, in load_module_py
    spec.loader.exec_module(module)
  File "<frozen importlib._bootstrap_external>", line 724, in exec_module
  File "<frozen importlib._bootstrap_external>", line 860, in get_code
  File "<frozen importlib._bootstrap_external>", line 791, in source_to_code
  File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
  File "/home/dev.internal/developer/sandbox/mbs/migrations/versions/9396eb21e075_tip_add_user_index_generated_user_.py", line 21
    op.add_column('usertable', sa.Column('user', sa.String(length=36), server_default=Computed(<sqlalchemy.sql.elements.TextClause object at 0x7febef2d30f0>, persisted=False), nullable=True))
```  

I found that that if I hacked in a __str__() implementation to the Computed class definition in SQLAlchemy I was able to use alembic successfully however I'm not sure if that's the most elegant solution.  

I'm using alembic version 1.3.0, sqlalchemy 1.3.11 (and MySQL 8.0.17 although I don't believe that matters as far as this issue is concerned).

I really appreciate any help or suggestions!

### LINKED ISSUE #628: Migration script wrongly generated when table name is all uppercase
I'm developing an API with Flask and I'm using flask-sqlachemy to create models and flask-migrate to generate migration scripts and upgrade the database. It is an Oracle database.

The first script ran smoothly and created the tables. But the next one where I changed a constraint on a column of a tablel wasn't created right. It is trying to create again all the tables and trying to delete the same tables with lowercase letters.

My guess is when I create tables with uppercase letters, the script transforms it to lowercase at somepoint and check if it exists as a lowercased string. The tables were created fine with all uppercase letters.

Model example:

`class Blacklist(db.Model):
''' IPs from users that continuously offend the system end up here '''

__bind_key__ = None
__tablename__ = 'SCR_BLACKLIST'

ID_BLACKED = db.Column(db.Integer, primary_key=True)
IP = db.Column(db.String(15), nullable=False)
BLACKLIST_INSERTION_TYPE = db.Column(
    db.Integer,
    db.ForeignKey('SCR_BLACKLIST_INSERTION_TYPES.ID_BLACK_TIPO'),
    nullable=False)
RATIO_LIMIT = db.Column(db.Integer, nullable=False, default=0)
IP_STATUS = db.Column(db.Boolean, nullable=False)
INSERTION_DATE = db.Column(db.DateTime, nullable=False, default=tb.now())
REMOVED_DATE = db.Column(db.DateTime, nullable=False)

def __repr__(self):
    print(f'<Blacklisted IP: {self.IP}>')`

Is there a way to workaround or solve this issue?

**OS:** Ubuntu 18.04
**Database:** Oracle
**Packages:** flask-sqlalchemy, flask-migrate

### PR DESCRIPTION
**Pull request by CaselIT**
Add alembic support for computed columns

This is still work in progress

Fix #624

### CODE REVIEW

**Comment by CaselIT** [comment]
I've made a typo in the branch name 😞

File: alembic/testing/requirements.py [hunk 1]
```diff
@@ -140,3 +140,14 @@ def comments_api(self):
     @property
     def alter_column(self):
         return exclusions.open()
+
+    @property
+    def db_computed_columns(self):
+        # TODO: should import the requirements from sqlalchemy
+        return exclusions.skip_if(["postgresql < 12", "sqlite", "mysql < 5.7"])
```
Comment (CaselIT) [inline alembic/testing/requirements.py]: @zzzeek here I should ideally return the value of `sqlalchemy.testing.requirements.SuiteRequirements.computed_columns` but currently it defines a `closed` condition. 
The real condition is set in `test.requirements.DefaultRequirements.computed_columns` which I cannot import here

Maybe the condition to indicate if a backend supports computed columns should be somewhere in the `sqlalchemy.testing` library so that it can be imported?

**Comment by CaselIT** [comment]
I think I'm stuck.
I'm not sure if it is possible to implement something that works reliably without inspect support for computed columns. I think that best outcome for now is to just ignore the computed column so it will not mess with the rest.

The current implementation in this pr works for (in pg and mysql):

- adding a table with a computed column
- adding a computed column to an existing table

The problems it has are (that I found):

- after adding a computed column the next migration without any change to that column will:
  - in mysql:
    ```py
    def upgrade():
        op.alter_column("square", "side", existing_type=mysql.FLOAT(), server_default="0", existing_nullable=True)
    
    def downgrade():
        op.alter_column("square", "side", existing_type=mysql.FLOAT(), server_default=sa.text("'0'"), existing_nullable=True)
    ```
  - in pg: 
    ```py
    def upgrade():
        op.alter_column(
            "square", "area", existing_type=postgresql.DOUBLE_PRECISION(precision=53), server_default=sa.Computed("side * side",), existing_nullable=True
        )    

    def downgrade():
        # This will raise error in pg when executed
        op.alter_column(
            "square", "area", existing_type=postgresql.DOUBLE_PRECISION(precision=53), server_default=sa.text("(side * side)"), existing_nullable=True
        )
    ```

- dropping a computed column will renders the wrong column in downgrade
  - in mysql:
  ```py
  def downgrade():
    op.add_column("square", sa.Column("area", mysql.FLOAT(), nullable=True))
  ```
  - in pg
  ```py
  def downgrade():
    # This will raise error in pg when executed
    op.add_column(
        "square",
        sa.Column("area", postgresql.DOUBLE_PRECISION(precision=53), server_default=sa.text("(side * side)"), autoincrement=False, nullable=True),
    )
  ```

In this branch https://github.com/CaselIT/alembic/tree/support_compute_changes I've tried to make the server default work with postgress, but it will break in mysql since its behaviour is different. I'm down think it is the right approach

**Comment by CaselIT** [comment]
found another issue

- changing a computed will be detected sometimes but it not work
  - in mysql:
  ```py
  def upgrade():
      op.alter_column("square", "side", existing_type=mysql.FLOAT(), server_default="0", existing_nullable=True)

  def downgrade():
      op.alter_column("square", "side", existing_type=mysql.FLOAT(), server_default=sa.text("'0'"), existing_nullable=True)
  ```
  - in pg
  ```py
  def upgrade():
    # This will raise error in pg when executed because there is not support for altering the computed
    op.alter_column(
        "square", "area", existing_type=postgresql.DOUBLE_PRECISION(precision=53), server_default=sa.Computed("side + side",), existing_nullable=True
    )

  def downgrade():
    # This will raise error in pg when executed
    op.alter_column(
        "square", "area", existing_type=postgresql.DOUBLE_PRECISION(precision=53), server_default=sa.text("(side * side)"), existing_nullable=True
    )
  ```

**Comment by zzzeek** [comment]
all the cases that would generate an "alter" should be ignored for now.   when the metadata column has a computed object on it, compare_server_default needs to return "no change", so that there is no false positive.

**Comment by CaselIT** [comment]
the downgrade error on pg will not be solved though

**Comment by zzzeek** [comment]
> the downgrade error on pg will not be solved though

OK, in that the server default is an intrinsic part of the column makes it a little more inconvenient than a separate constraint, however, if/when SQLAlchemy implements reflection for generated columns this part in Alembic will just start working, so it's not making things worse to implement rendering for starters.

in the world of database migrations, there's a few operations that are common, frequent, and important, and others that are less common, infrequent, not as important.

the important cases are:

1. create table
2. create index
3. alter table add column
4. alter table add constraint
5. adding ENUM members

most of which are detected right now and rendered correctly with the exception of a CHECK constraint being added and kind of critically the ENUM one, which is a long running quagmire kind of problem.

the non important cases are:

1. everything else - changing column type, changing column default, adding/removing column default, changing cols inside an index or constraint, changing index expressions, changing nullable, changing table name, changing column name, etc.    this is because these operations are not additive and break backwards compatibility so they are usually more awkward as the corresponding applications have to change as well, therefore they are less frequent and when they are undertaken, it's typical people have to edit their migration files anyway.

**Comment by CaselIT** [comment]
Ok @zzzeek, I trust your judgment on this.

I've updated the implementation to make alembic ignore changes when adding or removing computed to an existing column. It also ignores if the computed expression changes or the persistence changes.

It should be ready for review

There is still the error in postgresql when removing a computed column or computed from column, but I'm not sure currently there is a way to avoid it. I think a note in the documentation will have to suffice for now
I've made special tests for it, to test the current (erroneous) behaviour.

I've only tested on postgres and mysql, so I guess there may be some other surprises with other backends

To recap, in postgres the current implementation renders a erroneous downgrade (that will error if executed) when:
- removing a computed column
- removing computed from a column (without removing the column)

In both cases the problem is that the server default will be set to the expression that was in the computed class

**Comment by sqla-tester** [comment]
New Gerrit review created for change 7726d0957681634f9c772355463acb530d520462: https://gerrit.sqlalchemy.org/#/c/sqlalchemy/alembic/+/1600

**Comment by zzzeek** [comment]
OK ill start moving it in the queue.   we should absolutely try to implement reflection for GENERATED in SQLAlchemy also, it's just, implementing what we have so far in Alembic is not going to make things worse so we might as well.

**Comment by CaselIT** [comment]
I must have left a type somewhere. Also it seems that oracle behaves like pg

**Comment by CaselIT** [comment]
I think the new revision should pass in py2.7 and oracle

**Comment by zzzeek** [comment]
is this one ready for review?

**Comment by CaselIT** [comment]
I think so

**Comment by sqla-tester** [comment]
Patchset 16c760cb3fe6d55840d025bacabcb4188df0bc59 added to existing Gerrit review https://gerrit.sqlalchemy.org/#/c/sqlalchemy/alembic/+/1600

**Comment by CaselIT** [comment]
oh nice, it seems I'm able to reask to sqla-tester to run the tests once it has be assigned to a pull request.

**Comment by sqla-tester** [comment]
Patchset b1884461addb0c39a7d86ddc8e1139ab8ba1fa78 added to existing Gerrit review https://gerrit.sqlalchemy.org/#/c/sqlalchemy/alembic/+/1600

**Comment by sqla-tester** [comment]
Patchset 9c45651295626edf5f172ec827a5ced1440818cf added to existing Gerrit review https://gerrit.sqlalchemy.org/#/c/sqlalchemy/alembic/+/1600

**Comment by CaselIT** [comment]
all the tests should pass now. the previous build had only pep8 failures

**Comment by sqla-tester** [comment]
**mike bayer** (zzzeek) wrote:

hey....I think there is no DDL to actually alter a column that does not have a "generated" to have one, and vice versa.  I think you can only drop or add columns.    Which would render the whole "detection" issue sort of moot.

Looking at https://www.postgresql.org/docs/12/sql-altertable.html and trying everyhitng I can on the PG commandline, I can't add "GENERATED AS" to an existing column.    see https://dba.stackexchange.com/a/250869.

I'm working on this now and I think what I will do is that if changes to "computed" are detected with that server default thing, ignore it always, and if someone actually puts "Computed" in an op.alter_column() it has to raise.

**Comment by CaselIT** [comment]
That makes sense.
If I remember correctly the problem is that the generated column gets detected as a default in pg and oracle, so it's harder to discriminate it from an actual default value.

I don't know regarding the ddl if other backend support altering an existing column. Regarding pg they may be adding it in a later version, since iirc this version only had initial support for grnerated columns.

File: tests/test_autogen_computed.py [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+
+        diffs = self._fixture(m1, m2)
+
+        eq_(diffs[0][0], "remove_column")
+        eq_(diffs[0][2], "user")
+        c = diffs[0][3]
+        eq_(c.name, "foo")
+
+        assert c.computed is None
+        return c
```
Comment (sqla-tester) [inline tests/test_autogen_computed.py]: **Federico Caselli** (CaselIT) wrote:

nice we now have combinations also here

**Comment by CaselIT** [comment]
Maybe I could open an issue on sqlalchemy to track the inspection of computed columns

**Comment by sqla-tester** [comment]
**Federico Caselli** (CaselIT) wrote:

Thanks for the review Mike

**Comment by zzzeek** [comment]
> Maybe I could open an issue on sqlalchemy to track the inspection of computed columns

sure, the days of me keeping the number of open issues below 150 are long over .... :)

**Comment by CaselIT** [comment]
> sure, the days of me keeping the number of open issues below 150 are long over .... :)

I can wait :P

**Comment by sqla-tester** [comment]
**mike bayer** (zzzeek) wrote:

just a quick and easy drop in of Computed, no biggie....yeah right

**Comment by sqla-tester** [comment]
**Federico Caselli** (CaselIT) wrote:

> Patch Set 8:
> 
> just a quick and easy drop in of Computed, no biggie....yeah right

yes, this got quite a lot of bigger that the 4 lines required to render the computed string in add column :\

**Comment by CaselIT** [comment]
maybe we should just wait until sqlalchemy support inspecting the computed columns. 
Or at the very least do not report them as server_default in some backends?

**Comment by zzzeek** [comment]
> maybe we should just wait until sqlalchemy support inspecting the computed columns.
> Or at the very least do not report them as server_default in some backends?

i don't think it's really going to be that much of a problem.    putting in the render for Computed in a model is still essential whether or not detection works.

**Comment by zzzeek** [comment]
also a bunch of complexity here is because the combinations thing has proven to be so useful and I'm not ready to drop SQLA 1.2 support in alembic yet.

**Comment by CaselIT** [comment]
btw my suggestion was mainly regarding the workaround for server_default. Once it's not needed the code still needs to be there so support the sqlalchemy version that has computed but not the correct detection

**Comment by CaselIT** [comment]
Avoiding returning the default values seems very easy.

this is a patch for pg
```diff
diff --git a/lib/sqlalchemy/dialects/postgresql/base.py b/lib/sqlalchemy/dialects/postgresql/base.py
index 6d97033d0..1b0b2b46a 100644
--- a/lib/sqlalchemy/dialects/postgresql/base.py
+++ b/lib/sqlalchemy/dialects/postgresql/base.py
@@ -2827,7 +2827,14 @@ class PGDialect(default.DefaultDialect):
         table_oid = self.get_table_oid(
             connection, table_name, schema, info_cache=kw.get("info_cache")
         )
-        SQL_COLS = """
+
+        generated = (
+            "a.attgenerated as generated"
+            if self.server_version_info >= (12,)
+            else "'' as generated"
+        )
+        SQL_COLS = (
+            """
             SELECT a.attname,
               pg_catalog.format_type(a.atttypid, a.atttypmod),
               (SELECT pg_catalog.pg_get_expr(d.adbin, d.adrelid)
@@ -2836,7 +2843,8 @@ class PGDialect(default.DefaultDialect):
                AND a.atthasdef)
               AS DEFAULT,
               a.attnotnull, a.attnum, a.attrelid as table_oid,
-              pgd.description as comment
+              pgd.description as comment,
+              %s
             FROM pg_catalog.pg_attribute a
             LEFT JOIN pg_catalog.pg_description pgd ON (
                 pgd.objoid = a.attrelid AND pgd.objsubid = a.attnum)
@@ -2844,6 +2852,8 @@ class PGDialect(default.DefaultDialect):
             AND a.attnum > 0 AND NOT a.attisdropped
             ORDER BY a.attnum
         """
+            % generated
+        )
         s = (
             sql.text(SQL_COLS)
             .bindparams(sql.bindparam("table_oid", type_=sqltypes.Integer))
@@ -2876,6 +2886,7 @@ class PGDialect(default.DefaultDialect):
             attnum,
             table_oid,
             comment,
+            generated
         ) in rows:
             column_info = self._get_column_info(
                 name,
@@ -2886,6 +2897,7 @@ class PGDialect(default.DefaultDialect):
                 enums,
                 schema,
                 comment,
+                generated,
             )
             columns.append(column_info)
         return columns
@@ -2900,6 +2912,7 @@ class PGDialect(default.DefaultDialect):
         enums,
         schema,
         comment,
+        generated,
     ):
         def _handle_array_type(attype):
             return (
@@ -3010,6 +3023,11 @@ class PGDialect(default.DefaultDialect):
                 "Did not recognize type '%s' of column '%s'" % (attype, name)
             )
             coltype = sqltypes.NULLTYPE
+
+        # remove default from generated columns
+        if generated != '':
+            default = None
+
         # adjust the default value
         autoincrement = False
         if default is not None:
```

**Comment by CaselIT** [comment]
oracle (seems to require `all_tab_cols` instead of `all_tab_columns`)

```diff
diff --git a/lib/sqlalchemy/dialects/oracle/base.py b/lib/sqlalchemy/dialects/oracle/base.py
index fe3d586a4..d41f0ed34 100644
--- a/lib/sqlalchemy/dialects/oracle/base.py
+++ b/lib/sqlalchemy/dialects/oracle/base.py
@@ -1658,8 +1658,8 @@ class OracleDialect(default.DefaultDialect):
         text = """
             SELECT col.column_name, col.data_type, col.%(char_length_col)s,
               col.data_precision, col.data_scale, col.nullable,
-              col.data_default, com.comments\
-            FROM all_tab_columns%(dblink)s col
+              col.data_default, com.comments, col.virtual_column\
+            FROM all_tab_cols%(dblink)s col
             LEFT JOIN all_col_comments%(dblink)s com
             ON col.table_name = com.table_name
             AND col.column_name = com.column_name
@@ -1684,6 +1684,7 @@ class OracleDialect(default.DefaultDialect):
             nullable = row[5] == "Y"
             default = row[6]
             comment = row[7]
+            genetared = row[8]
 
             if coltype == "NUMBER":
                 if precision is None and scale == 0:
@@ -1708,6 +1709,9 @@ class OracleDialect(default.DefaultDialect):
                     )
                     coltype = sqltypes.NULLTYPE
 
+            if genetared == 'YES':
+                default = None
+
             cdict = {
                 "name": colname,
                 "type": coltype,
```

**Comment by CaselIT** [comment]
> hey....I think there is no DDL to actually alter a column that does not have a "generated" to have one, and vice versa. I think you can only drop or add columns. Which would render the whole "detection" issue sort of moot.

I've thought a bit about it and it may make sense that the expected way to modify a generated colum is to drop it and to regenerate it. Since it is generated it should usually be pretty safe to do*.
I'm not sure if alembic should do something like this though (I mean drop then re-add the column with the new generation)

* this in the backends that do not let the generated column expression contain other generated columns. From memory pg does not allow this, but mysql does. So the safeness of this operation may depend on the backend.

**Comment by CaselIT** [comment]
Maybe this logic (deop+readd) could be under a flag/configuration options (disabled by default)

But I'm not sure if this complexity is worth it

**Comment by zzzeek** [comment]
> Maybe this logic (drop+readd) could be under a flag/configuration options (disabled by default)
> 
> But I'm not sure if this complexity is worth it

not right now, I dont think this is going to be a widely used feature and autogenerate is not intended to be fully automatic.

**Comment by zzzeek** [comment]
it would help if I added one of the .py files to the gerrit

**Comment by CaselIT** [comment]
Maybe creating an external package for the testing stuff could be useful

**Comment by zzzeek** [comment]
the "combinations" thing was a huge change to the whole way the testing package works, usually things are not this out of sync.    I worry that having to put new releases of a separate testing package every time I need to fix something is going to make things more complicated, not to mention if the testing package itself relies upon SQLAlchemy features I'd have to be maintaining several branches of the testing package at the same time.

**Comment by CaselIT** [comment]
yes I agree that it can create some friction. maybe just the general framework that does not depend specifically on sqlalchemy could go there, like the catch warnings-errors, the assertions (eq_, is_, etc) and combinations.

but I guess the current situation is temporary since I guess alembic will not keep supporting version 1.1 and 1.2 forever

**Comment by zzzeek** [comment]
since the testing framework is also consumed by third party dialects and not just Alembic, having it be a separate package, probably one that has lots of logic and fallbacks to tell which SQLAlchemy version it runs against, is probably a good idea.   at the moment though I don't think I can handle it.

**Comment by zzzeek** [comment]
that is, rather than having lots of branches of the "testing framework" just have all the different SQLAlchemy assumptions present in one big release.

**Comment by CaselIT** [comment]
Let me know if there is anything else I can do for this pr

**Comment by sqla-tester** [comment]
**mike bayer** (zzzeek) wrote:

Code-Review+2 Workflow+1

**Comment by sqla-tester** [comment]
Gerrit review https://gerrit.sqlalchemy.org/1600 has been **merged**. Congratulations! :)
